### PR TITLE
[BEAM-10708] Read/Write Intermediate PCollections

### PR DESCRIPTION
**Pull request by KevinGG**
1. Added implementation of reading/writing cache directly on
   pipeline protos for intermediate PCollections.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>SDK</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by KevinGG** [comment]
R: @rohdesamuel 
PTAL, thx!

A test notebook that reads cache from a computed PCollection, applies a transform, and then writes the cache for the PCollection being computed.

etl_with_cache

**Comment by KevinGG** [comment]
Run Python_PVR_Flink PreCommit

File: sdks/python/apache_beam/runners/interactive/caching/write_cache.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  def write_cache(self) -> None:
+    """Writes cache for the cacheable PCollection that is being computed.
+    """
+    template, write_input_placeholder = self._build_runner_api_template()
+    input_placeholder_id = self._context.pcollections.get_id(
+        write_input_placeholder.placeholder_pcoll)
+    input_id = self._context.pcollections.get_id(self._cacheable.pcoll)
+
+    # Copy cache writing subgraph from the template to the pipeline proto.
+    for pcoll_id in template.components.pcollections:
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/interactive/caching/write_cache.py:63]: I'm not sure that this code is correct. The component ids (PCollection ids, coder ids, etc.) are specific to a given pipeline. In other words, given a component id in one pipeline there is no guarantee that the component is the same given the same component id in another pipeline.

Instead of doing this equality of ids between pipelines, it might be better to recursively iterate through the temporary pipeline and copy components over to _pipeline. You'll have to be careful though as the pipeline proto has some hidden ordering guarantees, like the transforms need to be in a topological ordering.

File: sdks/python/apache_beam/runners/interactive/augmented_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    if self._augmented:
+      return self
+    self._augmented = True
+
+    # Find pcolls eligible for reading or writing cache.
+    readcache_pcolls = set()
+    for pcoll, cacheable in self._cacheables.items():
+      key = repr(cacheable.to_key())
+      if (self._cache_manager.exists('full', key) and
+          pcoll in ie.current_env().computed_pcollections):
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/interactive/augmented_pipeline.py:117]: Is there ever a case where a PCollection is computed but doesn't exist?

File: sdks/python/apache_beam/runners/interactive/augmented_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    return c
+
+  def augment(self) -> 'AugmentedPipeline':
+    """Augments the pipeline with cache. Idempotent and returns self.
+
+    For a cacheable PCollection, if cache exists, read cache; else, write cache.
+    """
+    if self._augmented:
+      return self
+    self._augmented = True
```
Comment (rohdesamuel) [inline sdks/python/apache_beam/runners/interactive/augmented_pipeline.py:110]: Instead of having state that tracks whether the pipeline is augmented, it's a simpler API to have this always augment the user pipeline and let the client handle caching the result.

**Review by rohdesamuel — CHANGES_REQUESTED** [review]

File: sdks/python/apache_beam/runners/interactive/caching/write_cache.py (hunk 1 above)
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/caching/write_cache.py:63]: This I can explain: 
TL;DR: the `component_id_map` of a pipeline instance (also backs the pipeline context) is where each Python object is assigned a unique pipeline-scoped id. By making a shallow copy of it before augmenting and passing it into the temporary pipelines as a shared context for new components, we can guarantee all ids are unique even if they come from different pipeline instances.


- The underlying mechanism of the `component_id_map` are:
  -  a dictionary holds counters of different object types as `Dict[type, int]`;
  -  a dictionary holds object to id map as `Dict[Any, str]`.  

- All the temporary pipelines will generate components as if a sub graph in the original pipeline: ids of all the components are unique even to the current state of the augmented pipeline.

- The code moves the top level transforms of that subgraph into the runner api proto of the pipeline being augmented.

- The `pcoll_id in self._pipeline.components.pcollections ...` checks to skip ids from the shared context (`component_id_map`) that definitely do not exist in the temp pipeline and do not need copy at all.

- The `pcoll_id in write_input_placeholder.ignorable_components.pcollections ...` checks to skip unnecessary root transforms (because you don't need two root transforms) and placeholder subgraphs (as if they are pruned when we add cache transforms) since we only care about top level transforms that we need. It's kind of like moving everything there then pruning the unnecessary parts.

- The copy should preserve the topological ordering since it reads the proto of the temp pipeline in its original order and does the same thing to `set_proto_map` in `pipeline_instrument`.

File: sdks/python/apache_beam/runners/interactive/augmented_pipeline.py (hunk 2 above)
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/augmented_pipeline.py:117]: If the temp directory is cleaned up by other processes, or if we have a bug, then yes. I guess it doesn't hurt to check it here.

File: sdks/python/apache_beam/runners/interactive/augmented_pipeline.py (hunk 3 above)
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/augmented_pipeline.py:110]: That makes sense, I'll make a change to always calculate a new one.

File: sdks/python/apache_beam/runners/interactive/caching/read_cache.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      self._pipeline.components.pcollections[pcoll_id].CopyFrom(
+          template.components.pcollections[pcoll_id])
+    for coder_id in template.components.coders:
+      if coder_id in self._pipeline.components.coders:
+        continue
+      self._pipeline.components.coders[coder_id].CopyFrom(
+          template.components.coders[coder_id])
+    for windowing_strategy_id in template.components.windowing_strategies:
+      if windowing_strategy_id in \
+          self._pipeline.components.windowing_strategies:
```
[the remark this answers is not in the document]
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/caching/read_cache.py:73]: Thanks, I'll apply it to everywhere.

File: sdks/python/apache_beam/runners/interactive/caching/read_cache.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+      cache_manager: cache.CacheManager,
+      cacheable: Cacheable):
+    self._pipeline = pipeline
+    self._context = context
+    self._cache_manager = cache_manager
+    self._cacheable = cacheable
+    self._key = repr(cacheable.to_key())
+    self._label = '{}{}'.format('_cache_', self._key)
+
+  def read_cache(self) -> Tuple[str, str]:
```
[the remark this answers is not in the document]
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/caching/read_cache.py:52]: I'll add more comments.

File: sdks/python/apache_beam/runners/interactive/caching/write_cache.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+      cache_manager: cache.CacheManager,
+      cacheable: Cacheable):
+    self._pipeline = pipeline
+    self._context = context
+    self._cache_manager = cache_manager
+    self._cacheable = cacheable
+    self._key = repr(cacheable.to_key())
+    self._label = '{}{}'.format('_cache_', self._key)
+
+  def write_cache(self) -> None:
```
[the remark this answers is not in the document]
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/caching/write_cache.py:54]: Will do!

**Comment by KevinGG** [comment]
Run Python PreCommit

**Comment by KevinGG** [comment]
Rebased to take a flaky test fix.

**Comment by KevinGG** [comment]
Run Portable_Python PreCommit

**Review by rohdesamuel — APPROVED** [review]

**Comment by KevinGG** [comment]
The failure of action is from portable runner tests. Not related to this change.
